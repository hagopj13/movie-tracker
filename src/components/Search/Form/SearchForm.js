// @flow
import React, { useEffect } from 'react';
import { useForm } from 'react-hook-form';
import * as yup from 'yup';

import SearchField from 'components/Search/Field/SearchField';

type Props = {
  selectedQuery: string,
  onSubmit: ({ query: string }) => void,
};

const schema = yup.object().shape({
  query: yup.string().min(1).required(),
});

const SearchForm = (props: Props) => {
  const { selectedQuery, onSubmit } = props;

  const { register, handleSubmit, setValue } = useForm({
    mode: 'onSubmit',
    validationSchema: schema,
  });

  useEffect(() => {
    if (selectedQuery) {
      setValue('query', selectedQuery);
    }
  }, [selectedQuery, setValue]);

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <SearchField name="query" inputRef={register} />
    </form>
  );
};

export default SearchForm;
