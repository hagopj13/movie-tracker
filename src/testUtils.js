import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { createMount } from '@material-ui/core/test-utils';

import { initialState } from 'store/rootReducer';

const mount = createMount();
const mockStore = configureMockStore()(initialState);

export const mountWithStore = (children, store = mockStore) =>
  mount(<Provider store={store}>{children}</Provider>);
