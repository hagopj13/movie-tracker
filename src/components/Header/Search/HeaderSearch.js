// @flow
import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { useHistory, useLocation } from 'react-router-dom';

import searchActions from 'store/search/search.actions';
import searchSelectors from 'store/search/search.selectors';

import SearchForm from 'components/Search/Form/SearchForm';

type Props = {
  selectedSearchQuery: string,
  onSetSearchQuery: (query: string) => void,
};

export const HeaderSearch = (props: Props) => {
  const { selectedSearchQuery, onSetSearchQuery } = props;

  const history = useHistory();
  const location = useLocation();

  useEffect(() => {
    if (location.pathname === '/search') {
      const urlSearchParams = new URLSearchParams(location.search);
      const urlSearchQuery = urlSearchParams.get('query');
      if (urlSearchQuery && urlSearchQuery !== selectedSearchQuery) {
        onSetSearchQuery(urlSearchQuery);
      }
    }
  }, [location, selectedSearchQuery, onSetSearchQuery]);

  const handleSubmit = ({ query }) => {
    if (query) {
      history.push(`/search?query=${query}`);
    }
  };

  return <SearchForm onSubmit={handleSubmit} selectedQuery={selectedSearchQuery} />;
};

const mapStateToProps = createStructuredSelector({
  selectedSearchQuery: searchSelectors.selectQuery,
});

const mapDispatchToProps = {
  onSetSearchQuery: searchActions.setSearchQuery,
};

export default connect(mapStateToProps, mapDispatchToProps)(HeaderSearch);
