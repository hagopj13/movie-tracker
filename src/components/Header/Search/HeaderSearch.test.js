import React from 'react';
import { useLocation } from 'react-router-dom';
import { createShallow } from '@material-ui/core/test-utils';

import { mountWithStore } from 'testUtils';
import SearchForm from 'components/Search/Form/SearchForm';

import { HeaderSearch } from './HeaderSearch';

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: jest.fn(),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

describe('HeaderSearch component', () => {
  let shallow;
  let selectedSearchQuery;
  let wrapper;

  const mockSetSearchQuery = jest.fn();

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    jest.clearAllMocks();
    useLocation.mockReturnValue({ pathname: '/search', search: `?query=${selectedSearchQuery}` });
    selectedSearchQuery = 'Some query';
    const mockProps = {
      selectedSearchQuery,
      onSetSearchQuery: mockSetSearchQuery,
    };
    wrapper = shallow(<HeaderSearch {...mockProps} />);
  });

  it('should render the HeaderSearch component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should set the "selectedQuery" prop of the SearchForm component', () => {
    expect(wrapper.find(SearchForm).prop('selectedQuery')).toBe(selectedSearchQuery);
  });

  it('should push new query to history when SearchForm is submitted', () => {
    const newSearchQuery = 'Some new query';
    wrapper.find(SearchForm).prop('onSubmit')({ query: newSearchQuery });
    expect(mockHistoryPush).toHaveBeenLastCalledWith(`/search?query=${newSearchQuery}`);
  });

  it('should set the new search query after mounting', () => {
    const newSearchQuery = 'Some new query';
    useLocation.mockReturnValue({ pathname: '/search', search: `?query=${newSearchQuery}` });
    const mockProps = {
      selectedSearchQuery,
      onSetSearchQuery: mockSetSearchQuery,
    };
    mountWithStore(<HeaderSearch {...mockProps} />);
    expect(mockSetSearchQuery).toHaveBeenLastCalledWith(newSearchQuery);
  });

  it('should not set the search query after mounting if the query has not changed', () => {
    const mockProps = {
      selectedSearchQuery,
      onSetSearchQuery: mockSetSearchQuery,
    };
    mountWithStore(<HeaderSearch {...mockProps} />);
    expect(mockSetSearchQuery).not.toHaveBeenCalled();
  });
});
