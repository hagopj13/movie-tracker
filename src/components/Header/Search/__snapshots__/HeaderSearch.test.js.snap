// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderSearch component should render the HeaderSearch component correctly 1`] = `
<SearchForm
  onSubmit={[Function]}
  selectedQuery="Some query"
/>
`;
